fish-rust/src/wutil/wcstoi.rs

```diff
@@ -0,0 +1,221 @@
+
+/// Parse the given \p src as an integer.
+/// If mradix is not None, it is used as the radix; otherwise the radix is inferred:
+///   - Leading 0x or 0X means 16.
+///   - Leading 0 means 8.
+///   - Otherwise 10.
+/// The parse result contains the number as a u64, and whether it was negative.
+fn fish_parse_radix<Chars>(ichars: Chars, mradix: Option<u32>) -> Result<ParseResult, Error>
+where
+    Chars: Iterator<Item = char>,
+{
+    if let Some(r) = mradix {
+        assert!(
+            (2..=36).contains(&r),
+            "fish_parse_radix: invalid radix {}",
+            r
+        );
+    }
+    let chars = &mut ichars.peekable();
+
+    // Skip leading whitespace.
+    while current(chars).is_whitespace() {
+        chars.next();
+    }
+
+    if chars.peek().is_none() {
+        return Err(Error::Empty);
+    }
+
+    // Consume leading +/-.
+    let mut negative;
+    match current(chars) {
+        '-' | '+' => {
+            negative = current(chars) == '-';
+            chars.next();
+        }
+        _ => negative = false,
+    }
+
+    // Determine the radix.
+    let radix;
+    if mradix.is_some() {
+        radix = mradix.unwrap();
+    } else if current(chars) == '0' {
+        chars.next();
+        match current(chars) {
+            'x' | 'X' => {
+                chars.next();
+                radix = 16;
+            }
+            c if '0' <= c && c <= '9' => radix = 8,
+            _ => {
+                // Just a 0.
+                return Ok(ParseResult {
+                    result: 0,
+                    negative: false,
+                });
+            }
+        }
+    } else {
+        radix = 10;
+    }
+
+    // Compute as u64.
+    let mut consumed1 = false;
+    let mut result: u64 = 0;
+    while let Some(digit) = current(chars).to_digit(radix) {
+        result = result
+            .checked_mul(radix as u64)
+            .and_then(|r| r.checked_add(digit as u64))
+            .ok_or(Error::Overflow)?;
+        chars.next();
+        consumed1 = true;
+    }
+
+    // Did we consume at least one char?
+    if !consumed1 {
+        return Err(Error::InvalidDigit);
+    }
+
+    // Do not return -0.
+    if result == 0 {
+        negative = false;
+    }
+    Ok(ParseResult { result, negative })
+}
+
+/// Parse some iterator over Chars into some Integer type, optionally with a radix.
+fn fish_wcstoi_impl<Int, Chars>(src: Chars, mradix: Option<u32>) -> Result<Int, Error>
+where
+    Chars: Iterator<Item = char>,
+    Int: PrimInt,
+{
+    let bits = Int::zero().count_zeros();
+    assert!(bits <= 64, "fish_wcstoi: Int must be <= 64 bits");
+    let signed = Int::min_value() < Int::zero();
+
+    let ParseResult {
+        result, negative, ..
+    } = fish_parse_radix(src, mradix)?;
+
+    if !signed && negative {
+        Err(Error::InvalidDigit)
+    } else if !signed || !negative {
+        match Int::from(result) {
+            Some(r) => Ok(r),
+            None => Err(Error::Overflow),
+        }
```

Comment:
as turning `Option` into `Result` is a very common pattern in rust, you may use the method `.ok_or()` to avoid using the match statement:

```suggestion
        Int::from(result).ok_or(Error::Overflow)
```